### ISSUE TITLE
Better integration with streamly

### ISSUE THREAD

**Issue by ArulselvanMadhavan** [kowainik/co-log#142]
I'm trying to use the `co-log` library with `streamly` library and I haven't been able to figure out how to use the two together. Initially, I thought the problem was with my not being to implement `MonadTransControl` but it seems like there could be other ways to enable smooth integration of these two libraries.

https://github.com/kowainik/co-log/issues/134#issuecomment-520112368

Anyway, this issue is to enable integration with `streamly`. 
Thanks

**Comment by chshersh** [comment]
@ArulselvanMadhavan Could you provide some minimal code example of how you're using `co-log` with `streamly`? In general, `LoggerT` transformer is not the best approach for using `co-log`. I wonder if it's possible to use `co-log` in your case without relying on an additional `mtl`-style transformer monad.

I assume that you're using some code that involves `MonadAsync` which requires `MonadTransControl`. In general, I prefer to stay away from `MonadTransControl`, but I have no problems depending on `unliftio-core` and providing `MonadUnliftIO` constraint for `LoggerT` if this is the only way to use `co-log` with `streamly`.

See a similar issue in the `streamly` library. I've commented there as well:

* https://github.com/composewell/streamly/issues/63

cc @harendra-kumar

**Comment by harendra-kumar** [comment]
@ArulselvanMadhavan if you can provide some example code and the monad transformer stack that you are using for your use case, we can try to see if there is a way around without using MonadBaseControl.

**Comment by harendra-kumar** [comment]
@chshersh is there a fundamental issue in providing `MonadBaseControl` instance? You can avoid incorporating it in the library but I guess if it is possible @ArulselvanMadhavan can define his own instance to get it working if that is what is needed for him to proceed. In the long term, we can figure out if we can get rid of `MonadBaseControl` in streamly.

**Comment by chshersh** [comment]
@harendra-kumar I usually have hard times defining instances of `MonadBaseControl`. This abstraction has always been obscure to me, and the interface is not user-friendly. Also, `LoggerT` is not a trivial monad transformer, so it takes time for me to figure out lawful instance for it:

https://github.com/kowainik/co-log/blob/8dff51c38d2e2c8c2de5faae66938fada1b57d99/co-log/src/Colog/Monad.hs#L33-L35

Adding `MonadBaseControl` instance requires to add the `monad-control` library to the project dependencies. It is a common problem in Haskell, that if you want to provide reach interface, you end up depending on a lot of libraries. Each library can be small, but when you have a lot of them, your build times are increased significantly. I prefer to optimise dependencies footprint of my libraries. In this context, `MonadBaseControl` is a controversial abstraction, but `MonadUnliftIO` is not though it's less powerful. We use `MonadUnliftIO` in production, and it's considered one of the best-practices, so I'm okay to add `unliftio-core` package to dependencies. Also, `unliftio-core` is more lightweight because it depends only on `base` and one boot library.

**Comment by ArulselvanMadhavan** [comment]
I have a stream of files that I download and ingest into Redis in parallel. I'm trying to add logging support to this work/
```haskell
downloadFileAndIngest :: ( WithLog env String m, IsStream t) => t m Result
downloadFileAndIngest = 
   S.mapM readFromFile
$ S.mapM makeRedisConnection
$ S.map (\(fn, sn, _) -> (fn, sn)) -- Make RedisConnection
$ S.filter (validateFTPResult) -- Drop FTP Resultx
$ S.mapM downloadFileAndClose . S.mapM pickFileToDownload
```
I'm trying to use co-log in this way, in all stages of the streamly pipeline.

```haskell
pickFileToDownload :: WithLog env String m => m ()
pickFileToDownload = do
    logMsg "selecting file"
    fileToDownload <- selectFileToDownload
    logMsg $ "Selected file: " ++ show file
```

**Comment by chshersh** [comment]
@ArulselvanMadhavan Do you have your own custom monad or is working with `IO` is fine for you? If it's fine, then you can create `LogAction` separately (which you still need to do) and you can change your `pickFileToDownload` function to something like that:

```haskell
pickFileToDownload :: MonadIO m => LogAction m String -> m ()
pickFileToDownload log = do
    log <& "selecting file"
    fileToDownload <- selectFileToDownload
    log <& "Selected file: " ++ show file
```

Just pass `LogAction` explicitly, this simple solution should work without much problems. The logger is a value in `co-log` so you can do everything you can with the ordinary value. `WithLog` constraint is for custom monads when you don't want to pass `LogAction` explicitly to every function that requires logging.

**Comment by ArulselvanMadhavan** [comment]
I've been using IO and have resorted to passing `LogAction IO String` to methods as arguments. I was hoping that I'd be able to remove the explicit passing of `LogAction` using `LoggerT`

**Comment by chshersh** [comment]
@ArulselvanMadhavan if you want to stay in `IO` then passing `LogAction` explicitly is the only way to use `co-log`. `WithLog` constraint is for monads that have automatic access to `LogAction`: they somehow store it and can get/set it. For example, `ReaderT MyEnv IO` monad. If you use such a monad, then the best solution is to store `LogAction` inside `MyEnv`. With this approach, you can easily derive any instances for your `newtype` around `ReaderT MyEnv IO` because usually all monadic typeclasses have instances for `ReaderT` (including `MonadBaseControl`).

**Comment by harendra-kumar** [comment]
> results <- S.toList . parallely $ downloadFileAndIngest

@ArulselvanMadhavan just an aside, be aware that if you use `parallely` at the top level everything under it becomes parallel, this means all your `mapM` will become parallel. Quite often this is not desirable and can be counterproductive because of too much parallelism. So you could use `parallely` before the specific combinator you want to parallelize and then reset to serial after it using `serially`. In future we will try to have a better mechanism to make a single combinator parallel.

**Comment by chshersh** [comment]
I'm closing this issue for now. It looks like the primary problem in using `LogAction` was the absence of the `MonadBaseControl` instance when trying to use `logMsg` function directly with other `streamly` functions. If implementing this instance is still crucial to some applications, I propose to open a separate issue and discussion specific implementation under comments there. At the very end, `LogAction` is a value and you always can pass it as an argument and log via that value.
